src/librustc/middle/resolve_lifetime.rs

```diff
@@ -2066,27 +2070,78 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
         }
     }
 
+    fn lint_implicit_lifetimes_in_segment(&mut self,
+                                          segment_ident: ast::Ident,
+                                          generic_args: &'tcx hir::GenericArgs,
+                                          lifetime_refs: &[&'tcx hir::Lifetime]) {
+        let num_implicit_lifetimes = lifetime_refs.iter()
+            .filter(|lt| lt.name.is_implicit()).count();
+        if num_implicit_lifetimes == 0 {
+            return;
+        }
+
+        let mut err = self.tcx.struct_span_lint_node(
+            lint::builtin::ELIDED_LIFETIMES_IN_PATHS,
+            lifetime_refs[0].id, // FIXME: HirIdify #50928
+            segment_ident.span,
+            &format!("implicit lifetime parameters in types are deprecated"),
+        );
+
+        if num_implicit_lifetimes == 1 {
+            let (replace_span,
+                 suggestion) = if generic_args.args.len() == num_implicit_lifetimes &&
+                generic_args.bindings.is_empty() {
+                    // If there are no (non-implicit) generic args or bindings, our
+                    // suggestion includes the angle brackets
+                    (segment_ident.span.shrink_to_hi(), "<'_>")
+                } else {
+                    // Otherwise—sorry, this is kind of gross—we need to infer the
+                    // replacement point span from the generics that do exist
+                    let mut first_generic_span = None;
+                    for ref arg in &generic_args.args {
+                        match arg {
+                            hir::GenericArg::Lifetime(lt) => {
```

Comment:
I think there cannot be a lifetime name here... you can **totally** elide lifetimes (e.g., `Foo<T>`) but you cannot write only **some** of the needed lifetimes (e.g., `Foo<'_, T>` when `Foo<'_, '_, T>` is needed) -- you'll get an error in the latter case and (probably?) not go down this path. Should test though.